File: software/examples/delphi/ExampleCallback.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletGPS;

type
  TExample = class
  private
    ipcon: TIPConnection;
    gps: TBrickletGPS;
  public
    procedure CoordinatesCB(sender: TObject;
	                        const latitude: longword;
                            const ns: char;
                            const longitude: longword;
                            const ew: char;
                            const pdop: word;
                            const hdop: word;
                            const vdop: word;
                            const epe: word);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'ABC'; { Change to your UID }

var
  e: TExample;

{ Callback function for coordinates callback }
procedure TExample.CoordinatesCB(
sender: TObject;
const latitude: longword;
const ns: char;
const longitude: longword;
const ew: char;
const pdop: word;
const hdop: word;
const vdop: word;
const epe: word
);
begin
  WriteLn(Format('Latitude: %f ° %c', [latitude/1000000.0, ns]));
  WriteLn(Format('Longitude: %f ° %c', [longitude/1000000.0, ew]));
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  gps := TBrickletGPS.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }


  { Set Period for coordinates callback to 1s (1000ms)
    Note: The current callback is only called every second if the
          coordinates have changed since the last call! }
  gps.SetCoordinatesCallbackPeriod(1000);

  { Register coordinates callback to procedure CoordinatesCB }
  gps.OnCoordinates := {$ifdef FPC}@{$endif}CoordinatesCB;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy;
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExampleCallback.vb
Imports System
Imports Tinkerforge

Module ExampleCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your GPS Bricklet

    ' Callback subroutine for coordinates callback
    Sub CoordinatesCB(ByVal sender As BrickletGPS, ByVal latitude As Long, _
                      ByVal ns As Char, ByVal longitude As Long, ByVal ew As Char, _
                      ByVal pdop As Integer, ByVal hdop As Integer, _
                      ByVal vdop As Integer, ByVal epe As Integer)
        Console.WriteLine("Latitude: " + (latitude/1000000.0).ToString() + "° " + ns)
        Console.WriteLine("Longitude: " + (longitude/1000000.0).ToString() + "° " + ew)
        Console.WriteLine("")
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim gps As New BrickletGPS(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Register coordinates callback to subroutine CoordinatesCB
        AddHandler gps.Coordinates, AddressOf CoordinatesCB

        ' Set period for coordinates callback to 1s (1000ms)
        ' Note: The coordinates callback is only called every second
        '       if the coordinates has changed since the last call!
        gps.SetCoordinatesCallbackPeriod(1000)

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
